### ISSUE TITLE
Navigating to default screen from modal does nothing

### ISSUE THREAD

**Issue by chrisjenx** [hotwired/turbo-android#134]
Wasted a couple of hours on this, but it seems if you call: `navigate(non-modal context-uri)` from a fragment that was started with a modal context nothing happens - not just a it opens in a weird way, but literally nothing happens.

```
{
      "patterns": [
        "/login$", "/expert-login$"
      ],
      "properties": {
        "context": "modal",
        "uri": "turbo://fragment/login",
        "pull_to_refresh_enabled": false
      }
    },
    {
      "patterns": [
        "/passwords/new$"
      ],
      "properties": {
        "uri": "turbo://fragment/forgot-password",
        "fallback_uri": "turbo://fragment/web",
        "title": "Forgot Password",
        "pull_to_refresh_enabled": false
      }
    }
```

Calling forgot password from login fragment does nothing (Login is a NativeFragment fyi, and Forgot is a WebFragment).

Login was started as the starting location from the NavHostFragment

**Comment by chrisjenx** [comment]
I'm happy to fix once I've got some more time.

**Comment by jayohms** [comment]
Hey @chrisjenx, is that your complete configuration file? It's important to have a default configuration before more specific path configurations. An example is in the demo app: https://github.com/hotwired/turbo-android/blob/main/demo/src/main/assets/json/configuration.json#L5-L16

Have you [enabled debug logging](https://github.com/hotwired/turbo-android/blob/main/docs/ADVANCED-OPTIONS.md#enable-debug-logging)? This will output the navigation events and should clarify why a navigation isn't working properly. 

Happy to help further, but without the log or source, there's not enough information to go on.

**Comment by chrisjenx** [comment]
```
2021-01-25 10:58:10.053 5206-5328/com.zaarly.expert.debug V/RxViewModel: ----- view-effect NavigateForgot(forgetUrl=https://zweb-staging.zaarly.com/passwords/new, email=)
2021-01-25 10:58:10.054 5206-5206/com.zaarly.expert.debug D/TurboLog: shouldNavigateToLocation .......... [session: logged_out, location: https://zweb-staging.zaarly.com/passwords/new, shouldNavigate: true, fragment: LoginFragment]
2021-01-25 10:58:10.055 5206-5206/com.zaarly.expert.debug D/TurboLog: navigate .......................... [session: logged_out, location: https://zweb-staging.zaarly.com/passwords/new, options: TurboVisitOptions(action=ADVANCE, snapshotHTML=null, response=null), currentContext: MODAL, newContext: DEFAULT, presentation: PUSH, fragment: LoginFragment]
2021-01-25 10:58:10.056 5206-5206/com.zaarly.expert.debug D/TurboLog: dismissModalContextWithResult ..... [session: logged_out, location: https://zweb-staging.zaarly.com/passwords/new, uri: turbo://fragment/forgot-password, presentation: PUSH, fragment: LoginFragment]
```
```
{
  "settings": {
    "screenshots_enabled": false
  },
  "rules": [
    {
      "patterns": [
        ".*"
      ],
      "properties": {
        "uri": "turbo://fragment/web",
        "fallback_uri": "turbo://fragment/web",
        "pull_to_refresh_enabled": true
      }
    },
    {
      "patterns": [
        "/login$", "/expert-login$"
      ],
      "properties": {
        "context": "modal",
        "uri": "turbo://fragment/login",
        "presentation": "replace_root",
        "pull_to_refresh_enabled": false
      }
    },
    {
      "patterns": [
        "/inbox$", "/inbox/jobs$"
      ],
      "properties": {
        "uri": "turbo://activity/home",
        "pull_to_refresh_enabled": true
      }
    },
    {
      "patterns": [
        "/passwords/new$"
      ],
      "properties": {
        "uri": "turbo://fragment/forgot-password",
        "fallback_uri": "turbo://fragment/web",
        "presentation": "push",
        "pull_to_refresh_enabled": false
      }
    }
  ]
}
```
Log and config attached. When navigating to default from model it doesn't do anything... 
If I remove the model for default in the config for the login screen it works:

```
2021-01-25 11:00:51.867 5456-5577/com.zaarly.expert.debug V/RxViewModel: ----- view-effect NavigateForgot(forgetUrl=https://zweb-staging.zaarly.com/passwords/new, email=)
2021-01-25 11:00:51.873 5456-5456/com.zaarly.expert.debug D/TurboLog: shouldNavigateToLocation .......... [session: logged_out, location: https://zweb-staging.zaarly.com/passwords/new, shouldNavigate: true, fragment: LoginFragment]
2021-01-25 11:00:51.875 5456-5456/com.zaarly.expert.debug D/TurboLog: navigate .......................... [session: logged_out, location: https://zweb-staging.zaarly.com/passwords/new, options: TurboVisitOptions(action=ADVANCE, snapshotHTML=null, response=null), currentContext: DEFAULT, newContext: DEFAULT, presentation: PUSH, fragment: LoginFragment]
2021-01-25 11:00:51.875 5456-5456/com.zaarly.expert.debug D/TurboLog: navigateWithinContext ............. [session: logged_out, location: https://zweb-staging.zaarly.com/passwords/new, presentation: PUSH, fragment: LoginFragment]
2021-01-25 11:00:51.875 5456-5456/com.zaarly.expert.debug D/TurboLog: navigateToLocation ................ [session: logged_out, location: https://zweb-staging.zaarly.com/passwords/new, uri: turbo://fragment/forgot-password, fragment: LoginFragment]
2021-01-25 11:00:51.876 5456-5456/com.zaarly.expert.debug D/ForgotPasswordFragment: ForgotPassword: https://zweb-staging.zaarly.com/passwords/new
2021-01-25 11:00:51.887 5456-5456/com.zaarly.expert.debug D/TurboLog: fragment.onActivityCreated ........ [session: logged_out, location: https://zweb-staging.zaarly.com/passwords/new, fragment: ForgotPasswordFragment]
2021-01-25 11:00:51.887 5456-5456/com.zaarly.expert.debug D/TurboLog: fragment.onStart .................. [session: logged_out, location: https://zweb-staging.zaarly.com/passwords/new, fragment: ForgotPasswordFragment]
2021-01-25 11:00:51.889 5456-5456/com.zaarly.expert.debug D/TurboLog: fragment.onStop ................... [session: logged_out, location: https://zweb-staging.zaarly.com/expert-login, fragment: LoginFragment]
2021-01-25 11:00:51.907 5456-5456/com.zaarly.expert.debug D/TurboLog: visitLocationAsColdBoot ........... [session: logged_out, location: https://zweb-staging.zaarly.com/passwords/new]
```


Hope that helps

**Comment by jayohms** [comment]
Ah, I think I see what's going on. I think the problem exists here:
```json
{
  "patterns": [
    "/login$", "/expert-login$"
  ],
  "properties": {
    "context": "modal",
    "uri": "turbo://fragment/login",
    "presentation": "replace_root",
    "pull_to_refresh_enabled": false
  }
}
```

The intention of `modal` screens is that they overlay a `default` context. This helps manage the backstack in a number of navigation scenarios when the `modal` context exits (potentially across several Fragments) and returns to the `default` context. By using `"presentation": "replace_root"`, the root of the `NavHostFragment` is the modal Fragment and there's no existing underlying `default` context Fragment.

This is a situation that we haven't designed for. You could handle this in a couple different ways. If you're using the login fragment as the root, there shouldn't be a need to use a `modal` context through your login screens. If you used the following setup, after a successful login, you could navigate back to your home screen, which will become the new root of the `NavHostFragment`:
```json
{
  "patterns": [
    "/login$", "/expert-login$"
  ],
  "properties": {
    "uri": "turbo://fragment/login",
    "presentation": "replace_root",
    "pull_to_refresh_enabled": false
  }
},
{
  "patterns": [
    "/inbox$", "/inbox/jobs$"
  ],
  "properties": {
    "uri": "turbo://activity/home",
    "presentation": "replace_root",
    "pull_to_refresh_enabled": true
  }
}
```

I'll look into providing a log warning (or exception) for modal screens as the root Fragment, since I don't think that a scenario that we'll want to specifically handle.

If your login flow is fairly complex, or you have multiple entry points into your app, you may want to consider a multi-Activity setup. One Activity dedicated to the login flow and another dedicated to screens that require authentication. This is actually the setup we have in HEY, since our login flow has a variety of possible screens and keeping that flow segmented made sense. If you're interested in that setup, I'm happy to write up some documentation.

**Comment by chrisjenx** [comment]
Yeah I think throwing an error/exception would be fine, just making sure people see something happen vs silence :D

**Comment by jayohms** [comment]
@chrisjenx I've addressed this in https://github.com/hotwired/turbo-android/pull/136. A `TurboNavException` will now be thrown when replacing the root destination with a `modal` context destination. This will be available in the upcoming 7.0.0-beta05.

Thanks for your feedback 👍 . Closing this, but if something else doesn't appear to be working correctly, please let me know.

**Comment by jayohms** [comment]
`7.0.0-beta05` is available, which addresses this: 
https://github.com/hotwired/turbo-android/releases/tag/7.0.0-beta05
